v7/navstories/navstories.py

```diff
@@ -38,12 +38,50 @@ def set_site(self, site):
         site.scan_posts()
         # NAVIGATION_LINKS is a TranslatableSetting, values is an actual dict
         for lang in site.config['NAVIGATION_LINKS'].values:
+            # Which paths are navstories active for for lang?
+            navstories_paths = ()
+            if 'NAVSTORIES_PATHS' in site.config and lang in site.config['NAVSTORIES_PATHS']:
+                navstories_paths = site.config['NAVSTORIES_PATHS'][lang]
+
             new = []
+            newsub = {}
             for p in site.pages:
-                if lang in p.translated_to and not p.meta('hidefromnav'):
-                    new.append(p)
-            new_entries = tuple(sorted([(p.permalink(lang), p.title(lang)) for p in new]))
-
+                permalink = p.permalink(lang)
+                for s in navstories_paths:
+                    if permalink.startswith('/' + s + '/'):
+                        navpath = permalink[2 + len(s):].split('/') # Permalink format '/A/B/' for a story in s/A/B.rst
+                        if  navpath[-1] == '':
+                            del navpath[-1] # Also remove last element if empty
+                        if lang in p.translated_to and not p.meta('hidefromnav'):
+                            if len(navpath) <= 1:
+                                new.append(p)
+                            else:
+                                # Add key if not exists
+                                if not navpath[0] in newsub:
+                                    newsub[navpath[0]] = []
+                                # Add page to key
+                                newsub[navpath[0]].append((p.permalink(lang), p.title(lang)))
+            new_all = [(p.permalink(lang), p.title(lang)) for p in new]
+            for k in sorted(newsub.keys()):
+                new_all.append(tuple((tuple(newsub[k]), k)))
+            new_entries = []
+            navstories_mapping = ()
+            if 'NAVSTORIES_MAPPING' in site.config and lang in site.config['NAVSTORIES_MAPPING']:
+                navstories_mapping = site.config['NAVSTORIES_MAPPING'][lang]
+            for sk, sv in navstories_mapping:
+                for i in range(len(new_all)):
```

Comment:
1. You could use TranslatableSetting on all your setting to make things less of a mess.
2. Do the rename. And add some comments to explain your `range(len())` thing.
3. Aren’t you using `post.title()`? My point is, if you need that much customization for things (so page titles need to be different in nav and page), perhaps you’re better off making menus by hand, will be much simpler than messing with this plugin.